src/remote_callbacks.rs

```diff
@@ -450,3 +503,63 @@ extern "C" fn pack_progress_cb(
     })
     .unwrap_or(-1)
 }
+
+#[cfg(feature = "ssh")]
+extern "C" fn ssh_interactive_cb(
+    name: *const c_char,
+    name_len: c_int,
+    instruction: *const c_char,
+    instruction_len: c_int,
+    num_prompts: c_int,
+    prompts: *const raw::LIBSSH2_USERAUTH_KBDINT_PROMPT,
+    responses: *mut raw::LIBSSH2_USERAUTH_KBDINT_RESPONSE,
+    payload: *mut *mut c_void,
+) {
+    use libc::malloc;
+
+    panic::wrap(|| unsafe {
+        let prompts = prompts as *const libssh2_sys::LIBSSH2_USERAUTH_KBDINT_PROMPT;
+        let responses = responses as *mut libssh2_sys::LIBSSH2_USERAUTH_KBDINT_RESPONSE;
+
+        let name =
+            String::from_utf8_lossy(slice::from_raw_parts(name as *const u8, name_len as usize));
+        let instruction = String::from_utf8_lossy(slice::from_raw_parts(
+            instruction as *const u8,
+            instruction_len as usize,
+        ));
+
+        let mut wrapped_prompts = Vec::with_capacity(num_prompts as usize);
+        for i in 0..num_prompts {
+            let prompt = &*prompts.offset(i as isize);
+            wrapped_prompts.push(crate::cred::SshInteractivePrompt {
+                text: String::from_utf8_lossy(slice::from_raw_parts(
+                    prompt.text as *const u8,
+                    prompt.length as usize,
+                )),
+                echo: prompt.echo != 0,
+            });
+        }
+
+        let mut wrapped_responses = vec![String::new(); num_prompts as usize];
+
+        let payload = &mut *(payload as *mut Box<RemoteCallbacks<'_>>);
+        if let Some(callback) = &mut payload.ssh_interactive {
+            callback(
+                name.as_ref(),
+                instruction.as_ref(),
+                &wrapped_prompts[..],
+                &mut wrapped_responses[..],
+            );
+        }
+
+        for i in 0..num_prompts {
+            let response = &mut *responses.offset(i as isize);
+            let response_bytes = wrapped_responses[i as usize].as_bytes();
+
+            // libgit2 frees returned strings
+            let text = malloc(response_bytes.len());
+            response.text = text as *mut c_char;
+            response.length = response_bytes.len() as u32;
```

Comment:
There seems to be an issue here. The response is never copied into the buffer created via `malloc`. 

I think this also needs a check for a null return from `malloc`.

I'm only partially confident this is the correct way to handle this. From what I can tell, ssh2 has its own allocation routines that are part of the session, but I don't think the rust side exposes that, and I think it always defaults to libc malloc/free, so I don't think it is a problem. However, I'm a bit nervous about this manual memory handling with APIs that don't really document the behavior.